src/lib/broadcast.jl

```diff
@@ -232,23 +235,44 @@ end
 import ForwardDiff
 using ForwardDiff: Dual
 
-dual(x, p) = x
-dual(x::Real, p) = Dual(x, p)
-dual(x::Bool, p) = x
+
+# We do this because it ensures type stability so it compiles nicely on the gpu
+dual(x, i, N) = x
+dual(x::Bool, i, ::Val{N}) where {N} = x
+dual(x::Real, i, ::Val{N}) where {N} = Dual(x, ntuple(j-> i==j, Val(N)))
+# For complex since ForwardDiff.jl doesn't play nicely with complex numbers we
+# construct a Complex dual number and tag the real and imaginary parts separately
+function dual(x::Complex, i, ::Val{N}) where {N}
+    re_dual = Dual(real(x), ntuple(j->i==j, Val(2N)))
+    im_dual = Dual(imag(x), ntuple(j->(N+i)==j, Val(2N)))
```

Comment:
```suggestion
    re_dual = Dual(real(x), ntuple(==(i), 2N))
    im_dual = Dual(imag(x), ntuple(==(N+i), 2N))
```
Some local testing suggests that constant propagation should be able to figure this out without rewrapping with `Val`. If it doesn't though, feel free to disregard.

Reply:
Done!